### PR TITLE
TNPG: push all messages in one request and compress request payloads.

### PR DESCRIPTION
**Pull request by aforge**
Also, send pushes to `pushgw.tinode.co/push` by default.

### CODE REVIEW

File: server/push/tnpg/push_tnpg.go [hunk 1]
```diff
@@ -76,22 +96,19 @@ func sendPushes(rcpt *push.Receipt, config *configType) {
...
-			return
-		}
-		if code, status, err := postMessage(msg, config); err != nil {
-			log.Println("tnpg push failed:", err)
-			break
-		} else if code >= 300 {
-			log.Println("tnpg push rejected:", status, err)
-			break
-		}
+	messageMap := make(map[string]interface{})
```
[the remark this answers is not in the document]
Comment (aforge) [reply server/push/tnpg/push_tnpg.go]: Right. Thanks.

**Comment by or-else** [comment]
One consideration:

The pushgw gets the following message:
```
type messageData struct {
	Uid      t.Uid
	DeviceId string
	Message  *fcm.Message
}
```
But `Uid` and `DeviceId` it does not use or need. They are only used locally by push_fcm.go to remove invalid or expired tokens from the database. Pushgw is not going to do that. All it needs is `fcm.Message`.

If you decide to send the whole payload of `messageData`, it needs to be annotated with `json:"tag_tag"`.

Another consideration:
https://godoc.org/firebase.google.com/go/messaging#Message

For some inexplicable reason Google does not serialize some fields of the `Message`, like  `Message.Topic`:
```
Topic        string            `json:"-"`
```
I looked through the docs trying to figure out what is being dropped (there are quite a few of them in included structs) and seems like we don't use any of the dropped fields. But we do need to be aware of the problem and probably need to comment that some fields are not being sent.

**Comment by or-else** [comment]
I think the first item can be best addressed by changing the definition of `messageData`:
```
type messageData struct {
	fcm.Message
	Uid      t.Uid	`json:"-"`
	DeviceId string	`json:"-"`
}
```
And then it has to be referenced by pointer.

**Comment by aforge** [comment]
Okay, thanks. I've changed the code to send an array of fcm.Message only.

**Comment by or-else** [comment]
Something like this would avoid allocating and copying extra memory
```
    gz := gzip.NewWriter(w)
    json.NewEncoder(gz).Encode(data)
    gz.Close()
```

**Comment by aforge** [comment]
Thanks for suggestion. Done.

**Pull request merged**